From: Roman Gushchin <guro@fb.com>
Date: Mon, 23 Apr 2018 18:09:21 +0100
Subject: bpf: disable and restore preemption in __BPF_PROG_RUN_ARRAY
Patch-mainline: v4.17-rc4
Git-commit: 6899b32b5b2dee358936b82b8363b716607a138f
References: bsc#1109837

Running bpf programs requires disabled preemption,
however at least some* of the BPF_PROG_RUN_ARRAY users
do not follow this rule.

To fix this bug, and also to make it not happen in the future,
let's add explicit preemption disabling/re-enabling
to the __BPF_PROG_RUN_ARRAY code.

* for example:
 [   17.624472] RIP: 0010:__cgroup_bpf_run_filter_sk+0x1c4/0x1d0
 ...
 [   17.640890]  inet6_create+0x3eb/0x520
 [   17.641405]  __sock_create+0x242/0x340
 [   17.641939]  __sys_socket+0x57/0xe0
 [   17.642370]  ? trace_hardirqs_off_thunk+0x1a/0x1c
 [   17.642944]  SyS_socket+0xa/0x10
 [   17.643357]  do_syscall_64+0x79/0x220
 [   17.643879]  entry_SYSCALL_64_after_hwframe+0x42/0xb7

Signed-off-by: Roman Gushchin <guro@fb.com>
Cc: Alexei Starovoitov <ast@kernel.org>
Cc: Daniel Borkmann <daniel@iogearbox.net>
Acked-by: Alexei Starovoitov <ast@kernel.org>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Acked-by: Thomas Bogendoerfer <tbogendoerfer@suse.de>
---
 include/linux/bpf.h |    2 ++
 1 file changed, 2 insertions(+)

--- a/include/linux/bpf.h
+++ b/include/linux/bpf.h
@@ -337,11 +337,13 @@ int bpf_prog_array_copy_to_user(struct b
 	({						\
 		struct bpf_prog **_prog;		\
 		u32 _ret = 1;				\
+		preempt_disable();			\
 		rcu_read_lock();			\
 		_prog = rcu_dereference(array)->progs;	\
 		for (; *_prog; _prog++)			\
 			_ret &= func(*_prog, ctx);	\
 		rcu_read_unlock();			\
+		preempt_enable_no_resched();		\
 		_ret;					\
 	 })
 
